tools/cgpatch/src/CGPatchInstPass.cpp

```diff
@@ -0,0 +1,127 @@
+/**
+ * File: cgpatch-pass.cpp
+ * License: Part of the MetaCG project. Licensed under BSD 3 clause license. See LICENSE.txt file at
+ * https://github.com/tudasc/metacg/LICENSE.txt
+ */
+
+#include "nlohmann/json.hpp"
+#include "llvm/Demangle/Demangle.h"
+#include "llvm/IR/DerivedTypes.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/raw_ostream.h"
+#include <fstream>
+#include <iostream>
+#include <llvm/ADT/StringRef.h>
+#include <string>
+
+#include "CallAnalysis.h"
+#include "io/VersionTwoMCGWriter.h"
+
+using namespace llvm;
+using namespace cgpatch;
+
+// option parsing
+static cl::opt<bool> instrumentCtorsDtors("instrument-ctors-dtors", cl::desc("Instrument constructors and destructors"),
+                                          cl::init(false));
+
+static cl::opt<bool> verbose("cgpatch-verbose", cl::desc("Print debugging output"), cl::init(false));
+
+namespace {
+
+void insertMetaCGCall(Instruction& ins, Function& f, Value* calledOperand, Function* runtimeFunction);
+
+// Instrumentation
+void instrumentIndirectCalls(Module& M) {
+  ItaniumPartialDemangler demangler;
+  nlohmann::json j;
+
+  // Counter variables
+  int indirectCallCount{0};
+  int ctorDtorCallCount{0};
+  // Create function declaration
+  LLVMContext& Context = M.getContext();
+  auto functionType =
+      FunctionType::get(Type::getVoidTy(Context), {PointerType::get(Context, 0), PointerType::get(Context, 0)}, false);
+  Function* runtimeFunction = cast<Function>(M.getOrInsertFunction("__metacg_indirect_call", functionType).getCallee());
+
+  for (Function& F : M) {
+    if (verbose) {
+      outs() << "Traversing function: " << F.getName() << "\n";
+    }
+    for (BasicBlock& B : F)
+      for (Instruction& Ins : B) {
+        // Check if Ins is a call instruction
+        auto* CB = dyn_cast<CallBase>(&Ins);
+        auto CT = detectCallType(CB);
+
+        if (verbose) {  // Only print if verbose is true
+          printCallTypeInfo(CT, CB);
+        }
+        if (CT == CallType::Unknown)
+          continue;
+
+        auto calledFunction = CB->getCalledFunction();
```

Comment:
I think this should be moved inside the `if` body, as it is not used anywhere else FWICT